### PR TITLE
Simplify serialization protocols

### PR DESCRIPTION
**Pull request by vyasr**
## Description

This rewrites all serialization protocols in cudf to remove the need for pickling intermediates.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Review by Matt711 — COMMENTED** [review]
Hey @vyasr, are you going to retarget to 25.02?

**Comment by vyasr** [comment]
No, this PR is intended for 24.12 (in spite of code freeze).

File: python/cudf/cudf/core/dtypes.py [hunk 1]
```diff
@@ -91,13 +90,13 @@ def dtype(arbitrary):
         raise TypeError(f"Cannot interpret {arbitrary} as a valid cuDF dtype")
 
 
-def _decode_type(
+def _check_type(
```
Comment (Matt711) [inline python/cudf/cudf/core/dtypes.py:93]:
```suggestion
def _validate_metadata(
```

File: python/cudf/cudf/core/dtypes.py [hunk 2]
```diff
@@ -91,13 +90,13 @@ def dtype(arbitrary):
...
     cls: type,
     header: dict,
     frames: list,
     is_valid_class: Callable[[type, type], bool] = operator.is_,
-) -> tuple[dict, list, type]:
-    """Decode metadata-encoded type and check validity
+) -> None:
+    """Perform metadata-encoded type and check validity
```
Comment (Matt711) [inline python/cudf/cudf/core/dtypes.py:99]:
```suggestion
    """Validate the deserialization metadata
```

File: python/cudf/cudf/core/dtypes.py [hunk 3]
```diff
@@ -305,29 +297,29 @@ def construct_from_string(self):
...
+            )
         header["categories"] = categories_header
         frames.extend(categories_frames)
         header["frame_count"] = len(frames)
         return header, frames
 
     @classmethod
     def deserialize(cls, header, frames):
-        header, frames, klass = _decode_type(cls, header, frames)
+        _check_type(cls, header, frames)
```
Comment (Matt711) [inline python/cudf/cudf/core/dtypes.py:315]: And others
```suggestion
        _validate_metadata(cls, header, frames)
```

**Review by Matt711 — APPROVED** [review]
Thanks, the changes look good to me (other a couple non-blocking suggestions). I paid closer attention to `Serializable` and `Frame` changes, since the other changes, for the most part remove pickle calls.

**Comment by vyasr** [comment]
@Matt711 thanks for your suggestions! Since they are non-blocking improvements and this PR is time-sensitive, I'm going to go ahead and get this merged and then work on those in a follow-up on 25.02.

**Pull request merged**